go/vt/vtgate/planbuilder/operators/insert.go

```diff
@@ -111,12 +111,228 @@ func createOperatorFromInsert(ctx *plancontext.PlanningContext, ins *sqlparser.I
+		// this needs a delete before insert as there can be row clash which needs to be deleted first.
+		ins.Action = sqlparser.InsertAct
+		deleteBeforeInsert = true
+	}
+
+	insOp, err := checkAndCreateInsertOperator(ctx, ins, vTbl, routing)
+	if err != nil {
+		return nil, err
+	}
+
+	if !deleteBeforeInsert {
+		return insOp, nil
+	}
+
+	rows, isRows := ins.Rows.(sqlparser.Values)
+	if !isRows {
+		return nil, vterrors.VT12001("REPLACE INTO using select statement")
+	}
+
+	pkCompExpr := pkCompExpression(vTbl, ins, rows)
+	uniqKeyCompExprs, err := uniqKeyCompExpressions(vTbl, ins, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	whereExpr := getWhereCondExpr(append(uniqKeyCompExprs, pkCompExpr))
+
+	delStmt := &sqlparser.Delete{
+		TableExprs: sqlparser.TableExprs{sqlparser.CloneRefOfAliasedTableExpr(ins.Table)},
+		Where:      sqlparser.NewWhere(sqlparser.WhereClause, whereExpr),
+	}
+	delOp, err := createOpFromStmt(ctx, delStmt, false, "")
+	if err != nil {
+		return nil, err
+	}
+	return &Sequential{Sources: []ops.Operator{delOp, insOp}}, nil
+}
+
+func getWhereCondExpr(compExprs []*sqlparser.ComparisonExpr) sqlparser.Expr {
+	var outputExpr sqlparser.Expr
+	for _, expr := range compExprs {
+		if expr == nil {
+			continue
+		}
+		if outputExpr == nil {
+			outputExpr = expr
+			continue
+		}
+		outputExpr = &sqlparser.OrExpr{
+			Left:  outputExpr,
+			Right: expr,
+		}
+	}
+	return outputExpr
+}
+
+func pkCompExpression(vTbl *vindexes.Table, ins *sqlparser.Insert, rows sqlparser.Values) *sqlparser.ComparisonExpr {
+	if len(vTbl.PrimaryKey) == 0 {
+		return nil
+	}
+	type pComp struct {
+		idx int
+		def sqlparser.Expr
+	}
+	var pIndexes []pComp
+	var pColTuple sqlparser.ValTuple
+	for _, pCol := range vTbl.PrimaryKey {
+		var def sqlparser.Expr
+		idx := ins.Columns.FindColumn(pCol)
+		if idx == -1 {
+			def = findDefault(vTbl, pCol)
+			if def == nil {
+				// If default value is empty, nothing to compare as it will always be false.
+				return nil
+			}
+		}
+		pIndexes = append(pIndexes, pComp{idx, def})
+		pColTuple = append(pColTuple, sqlparser.NewColName(pCol.String()))
+	}
+
+	var pValTuple sqlparser.ValTuple
+	for _, row := range rows {
+		var rowTuple sqlparser.ValTuple
+		for _, pIdx := range pIndexes {
+			if pIdx.idx == -1 {
+				rowTuple = append(rowTuple, pIdx.def)
+			} else {
+				rowTuple = append(rowTuple, row[pIdx.idx])
+			}
+		}
+		pValTuple = append(pValTuple, rowTuple)
+	}
+	return sqlparser.NewComparisonExpr(sqlparser.InOp, pColTuple, pValTuple, nil)
+}
+
+func findDefault(vTbl *vindexes.Table, pCol sqlparser.IdentifierCI) sqlparser.Expr {
+	for _, column := range vTbl.Columns {
+		if column.Name.Equal(pCol) {
+			return column.Default
+		}
+	}
+	panic(vterrors.VT03014(pCol.String(), vTbl.Name.String()))
+}
+
+type uComp struct {
+	idx int
+	def sqlparser.Expr
+}
+
+func uniqKeyCompExpressions(vTbl *vindexes.Table, ins *sqlparser.Insert, rows sqlparser.Values) (comps []*sqlparser.ComparisonExpr, err error) {
```

Comment:
we can use `panic(err)` instead of returning errors in the operator package

Reply:
I will change in followup PR